Mathlib/Topology/Algebra/IsOpenUnits.lean

```diff
@@ -0,0 +1,90 @@
+/-
+Copyright (c) 2025 Andrew Yang. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Andrew Yang
+-/
+import Mathlib.RingTheory.Jacobson.Ideal
+import Mathlib.Topology.Algebra.GroupWithZero
+import Mathlib.Topology.Algebra.Nonarchimedean.AdicTopology
+
+/-!
+
+# Topological monoids with open units
+
+We say that a topological monoid `M` has open units (`IsOpenUnits`) if `Mˣ` is open in `M` and
+has the subspace topology (i.e. inverse is continuous).
+
+Typical examples include monoids with discrete topology, topological groups (or fields),
+and rings `R` equipped with the `I`-adic topology where `I ≤ J(R)` (`IsOpenUnits.of_isAdic`).
+
+This condition is needed so that for an affine scheme `X` over `R` and `U ⊆ X` an affine subscheme,
+`U(R)` is an open subspace of `X(R)`.
+-/
+
+open Topology
+
+/--
+We say that a topological monoid `M` has open units if `Mˣ` is open in `M` and
+has the subspace topology (i.e. inverse is continuous).
+
+Typical examples include monoids with discrete topology, topological groups (or fields),
+and rings `R` equipped with the `I`-adic topology where `I ≤ J(R)`.
+-/
+@[mk_iff]
+class IsOpenUnits (M : Type*) [Monoid M] [TopologicalSpace M] : Prop where
+  isOpenEmbedding_unitsVal : IsOpenEmbedding (Units.val : Mˣ → M)
+
+instance (priority := 900) (M : Type*) [Monoid M] [TopologicalSpace M] [DiscreteTopology M] :
+    IsOpenUnits M := ⟨.of_continuous_injective_isOpenMap Units.continuous_val Units.ext
+      fun _ _ ↦ isOpen_discrete _⟩
+
+instance (priority := 900) {M : Type*} [Group M] [TopologicalSpace M] [ContinuousInv M] :
+    IsOpenUnits M where
+  isOpenEmbedding_unitsVal := toUnits_homeomorph.symm.isOpenEmbedding
+
+instance (priority := 900) {M : Type*} [GroupWithZero M]
+    [TopologicalSpace M] [HasContinuousInv₀ M] [T1Space M] : IsOpenUnits M where
```

Comment:
So the `p`-adic rationals automatically acquire an instance, but the `p`-adic integers do not (yet), right?